From cfb6aacb02418681a792025e7d461389de5d704a Mon Sep 17 00:00:00 2001
From: Konrad Leszczynski <konrad.leszczynski@intel.com>
Date: Tue, 15 Dec 2015 11:28:28 +0000
Subject: [PATCH 02/12] BACKPORT: usb: gadget: f_midi: check kstrdup() return
 value

kstrdup() might fail, so check its return value and react appropriately.

Change-Id: I4573316903c38146f1e2b429fc7b2d2b6f4fd17c
Tracked-On: https://jira01.devtools.intel.com/browse/OAM-3355
Signed-off-by: Andrzej Pietrasiewicz <andrzej.p@samsung.com>
Signed-off-by: Felipe Balbi <balbi@ti.com>
Signed-off-by: Konrad Leszczynski <konrad.leszczynski@intel.com>
Reviewed-on: https://android.intel.com:443/448828
---
 drivers/usb/gadget/f_midi.c | 23 ++++++++++++++++++-----
 1 file changed, 18 insertions(+), 5 deletions(-)

diff --git a/drivers/usb/gadget/f_midi.c b/drivers/usb/gadget/f_midi.c
index 4b8af19..67994db 100644
--- a/drivers/usb/gadget/f_midi.c
+++ b/drivers/usb/gadget/f_midi.c
@@ -654,6 +654,14 @@ static struct snd_rawmidi_ops gmidi_out_ops = {
 	.trigger = f_midi_out_trigger
 };
 
+static inline void f_midi_unregister_card(struct f_midi *midi)
+{
+	if (midi->card) {
+		snd_card_free(midi->card);
+		midi->card = NULL;
+	}
+}
+
 /* register as a sound "card" */
 static int f_midi_register_card(struct f_midi *midi)
 {
@@ -716,10 +724,7 @@ static int f_midi_register_card(struct f_midi *midi)
 	return 0;
 
 fail:
-	if (midi->card) {
-		snd_card_free(midi->card);
-		midi->card = NULL;
-	}
+	f_midi_unregister_card(midi);
 	return err;
 }
 
@@ -968,15 +973,23 @@ int __init f_midi_bind_config(struct usb_configuration *c,
 	midi->func.disable     = f_midi_disable;
 
 	midi->id = kstrdup(id, GFP_KERNEL);
+	if (id && !midi->id) {
+		status = -ENOMEM;
+		goto kstrdup_fail;
+	}
 	midi->buflen = buflen;
 	midi->qlen = qlen;
 
 	status = usb_add_function(c, &midi->func);
 	if (status)
-		goto setup_fail;
+		goto add_fail;
 
 	return 0;
 
+add_fail:
+	kfree(midi->id);
+kstrdup_fail:
+	f_midi_unregister_card(midi);
 setup_fail:
 	for (--i; i >= 0; i--)
 		kfree(midi->in_port[i]);
-- 
1.9.1

